### PR TITLE
Write app files

### PR DESCRIPTION
**Pull request by jamie-lynch**
## What does this change?

This PR updates the functionality of existing commands to allow them to also write out the required app files. To add this new functionality, a number of changes were required:

* The `CDKBuilder` class was renamed `StackBuilder` and moved to the `utils/stacks.ts` file
* A new `AppBuilder` class was added
* The arguments taken by the commands were modified to include app details and the path to the CDK directory rather than specific file
* "Basic" imports (e.g. `import "test"`) were added
* Custom sorting logic was adding to the imports rendering to ensure imports are rendered in the correct order
* The rendering of imports was moved to the `Imports` class to remove duplication
* Two new factory functions were added to the `utils/imports.ts` file to create `Imports` classes with the correct default imports for apps and stacks
* A codemaker test util was added to remove duplication (and the tsconfig was updated to account for the new `./test` directory)

## How to test

Run the commands and observe the new functionality.

## How can we measure success?

More of the required boilerplate is generated automagically by the CDK.

### CODE REVIEW

File: src/commands/init.ts [hunk 1]
```diff
@@ -46,11 +46,15 @@ export class InitCommand extends Command {
...
-    checkDirectoryIsEmpty(config.outputDir);
+    InitCommand.validateConfig(config);
 
     return config;
   };
 
+  static validateConfig = (config: InitCommandConfig): void => {
+    checkDirectoryIsEmpty(config.outputDir);
+  };
+
```
Comment (stephengeller) [inline src/commands/init.ts:57]: Is this refactor to allow us to perform more actions to validate config in the future?

**Comment by stephengeller** [comment]
Tried running this locally on the branch and the commands aren't showing up when running `./bin/run`, can you double-check it works for you?

**Comment by stephengeller** [comment]
```
projects/guardian/cdk-cli ‹jl/write-app-files› % ./bin/run
A CLI tool to migrate from cloudformation to Guardian flavoured CDK

VERSION
  @guardian/cdk-cli/0.0.0 darwin-x64 node-v14.15.1

USAGE
  $ cdk-cli [COMMAND]

COMMANDS
  help  display help for cdk-cli
```

File: src/commands/init.ts (hunk 1 above)
Comment (jamie-lynch) [reply src/commands/init.ts:57]: In part, yes. It separates the concerns rather than doing both in `getConfig` which hopefully improves the readability and also makes it easier to test.

**Comment by jamie-lynch** [comment]
> Tried running this locally on the branch and the commands aren't showing up when running `./bin/run`, can you double-check it works for you?

Ah yes. I think you now have to build before you can use `./bin/run`. There's a `./script/build` script to do that. 

I don't remember this being a requirement before so maybe it's something that changed with an update to one of the `oclif` libraries. One for me to look into further.

**Comment by stephengeller** [comment]
Gotcha, thanks. Ran `new` and it seemed to work nicely (although I had to create the dir first), but ran into this issue when running `init`:

```
projects/guardian/cdk-cli ‹jl/write-app-files› % ./bin/run init                                                                                                                                                  1 ↵
Starting CDK generator
Creating ./cdk
Copying template files
    Error: ENOENT: no such file or directory, scandir '/Users/stephen_geller/projects/guardian/cdk-cli/lib/commands/../template'
    Code: ENOENT
```

**Comment by jamie-lynch** [comment]
> Gotcha, thanks. Ran `new` and it seemed to work nicely (although I had to create the dir first), but ran into this issue when running `init`:
> 
> ```
> projects/guardian/cdk-cli ‹jl/write-app-files› % ./bin/run init                                                                                                                                                  1 ↵
> Starting CDK generator
> Creating ./cdk
> Copying template files
>     Error: ENOENT: no such file or directory, scandir '/Users/stephen_geller/projects/guardian/cdk-cli/lib/commands/../template'
>     Code: ENOENT
> ```
> 
> Can pick this up tomorrow and pair on it as I expect it's something to do with my local environment!

Ah - it also doesn't copy the `template` directory over when it runs the build. Presumably because those files are all static. Sorry, I'd forgotten about that. Another thing for me to fix.

File: src/utils/app.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      this.code.line();
+    }
+
+    this.config.imports.render(this.code);
+
+    this.code.line("const app = new App();");
+
+    this.config.stacks.forEach((stack) => {
+      this.code.line(
+        `new ${stack.name}(app, "${stack.name}", { app: "${kebabCase(
```
Comment (jacobwinch) [inline src/utils/app.ts:48]: I think we may need to convert `stack.name` to upper camel case here?

File: src/utils/stack.ts [hunk 3]
```diff
@@ -7,46 +7,49 @@ interface CdkParameterProps extends CfnParameterProps {
...
+    if (this.config.comment) {
+      this.code.line(this.config.comment);
+      this.code.line();
+    }
 
-    this.addImports();
+    this.config.imports.render(this.code);
 
     this.code.openBlock(
       `export class ${this.config.stackName} extends GuStack`
```
Comment (jacobwinch) [inline src/utils/stack.ts:49]: I think this is another case where we want to upper camel case the stack name?

File: src/utils/app.ts (hunk 2 above)
Comment (jamie-lynch) [reply src/utils/app.ts:48]: Ah yes. From what I remember, we make the assumption that the stack name provided is already in upper camel case but it would be better to convert it. I was already tempted to include both the upper camel case and kebab case values in the config object to avoid repeatedly converting so maybe that change is worth making. Do you think it's better to make that change in this PR or in a follow up?

Comment (jacobwinch) [reply src/utils/app.ts:48]: Perhaps something for a follow-up PR; I just happened to test this PR out using an edge case stack name, so thought I'd mention it!

**Review by jacobwinch — APPROVED** [review]
I've tested these commands out and it's working well for me 🎉 

I ran into a few problems, but they are either minor (https://github.com/guardian/cdk-cli/pull/38#discussion_r566697145) or addressed in different PRs (https://github.com/guardian/cdk-cli/pull/60 and https://github.com/guardian/cdk-cli/pull/49), so let's get this merged in!

Great work @jamie-lynch 💯
